homeassistant/components/godice/__init__.py

```diff
@@ -0,0 +1,80 @@
+"""The godice integration."""
+
+from __future__ import annotations
+
+import logging
+
+import bleak
+import godice
+
+from homeassistant.components import bluetooth
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.device_registry import DeviceInfo
+
+from .const import DOMAIN
+
+PLATFORMS = [Platform.SENSOR]
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up godice from a config entry."""
+    _LOGGER.debug("Setup started")
+
+    def on_disconnect_callback(_ble_data):
+        _LOGGER.debug("on_disconnect_callback called")
+        is_disconnected_by_request = hass.data[DOMAIN][entry.entry_id][
+            "is_disconnected_by_request"
+        ]
+        if not is_disconnected_by_request:
+            hass.create_task(hass.config_entries.async_reload(entry.entry_id))
+
+    ble_device = bluetooth.async_ble_device_from_address(hass, entry.data["address"])
+    assert ble_device is not None
```

Comment:
I wouldn't use an assertion here. What if some BLEDevice becomes available shortly after async_setup_entry is called? Then it will only become available once HA restarts again. 

You can also register a callback with HA in your async_setup_entry that will update the ble_device for your dice. This is useful if your device becomes available through a different adapter or bluetooth proxy:
```python
from homeassistant.components.bluetooth import (
    BluetoothCallbackMatcher,
    BluetoothChange,
    BluetoothScanningMode,
    BluetoothServiceInfoBleak,
    async_ble_device_from_address,
    async_register_callback,
)
from homeassistant.core import callback
    @callback
    def async_update_ble_device(
        service_info: BluetoothServiceInfoBleak, change: BluetoothChange
    ) -> None:
        """Update the BLEDevice."""
        dice.update_ble_device(service_info.device, rssi=service_info.advertisement.rssi)
```

Reply:
initially I used `assert not None` to fix type check failure - `async_ble_device_from_address` returns BLEDevice | None, while `BleakClient` first arg is BLEDevice|str. But yes I agree that in case `ble_device` is None then AssertionError will be raised and HA will skip connection retry. I've put that code snipped under try-catch too to make sure I wrap an exception with `ConfigEntryNotReady` if any

Reply:
If you ever get a type check failure you can also use TYPE_CHECKING constant which will only execute the assertion for type checkers:
```python
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    assert ble_device is not None
```

But in this case I think the best way to solve it is indeed to use `ConfigEntryNotReady`, then HA will automatically retry it later. I would still add the `async_register_callback()` though to always have access to the latest BLEDevice

Reply:
does not `async_ble_device_from_address` return the latest ble_device ? I supposed it does. Theoretically I can register a callback to receive the last ble_device too after a call to `async_ble_device_from_address` but I'm not sure I need it anywhere now. I try to connect once when the integration loads, then if it fails I just raise `ConfigEntryNotReady` leverage HA reconnect. In case of receiving new `ble_device` while integration is alive I will just lose it after initiating integration reload since as I understand all vars are lost between reloads.
Please correct me if I'm wrong

Reply:
Do you only connect once when the integration loads and not anymore after that? In that case I don't think you need the latest BLEDevice, but if the user can initiate a connection themselves then you should use the latest BLEDevice for that as it contains information on what adapter it is available on

Reply:
Correct, I connect once only when the integration loads